File: framework/languages/node/e2e/RegistrationCodec/Client/Scenarios/RcB2ProtobufCodecScenario.ts
// RC-B2: Root에 등록한 Protobuf extension을 사용한다 시나리오를 검증한다.
import type { CodecScenarioRes } from '../../Shared/messages';
import { postJson } from '../../../http-client';
import { ensure } from '../Support/scenario-assert';

// RC-B2 verifies Protobuf content-type selection on the main host's global codec registry.
export async function runRcB2(serverUrl: string): Promise<void> {
  const reply = await postJson<CodecScenarioRes>(serverUrl, '/codec/protobuf');
  ensure(reply.value === 'echo:rc-b2', 'RC-B2 reply value mismatch.');
  ensure(reply.contentType === 'application/x-protobuf', 'RC-B2 expected Protobuf content type.');
  const evidence = await postJson<readonly string[]>(serverUrl, '/evidence/wait', {
    containsAll: ['codec=protobuf', 'rc-b2-send', 'codec-reply|codec=protobuf'],
    timeoutMilliseconds: 10_000
  });
  ensure(evidence.some((line) => line.includes('codec-request|codec=protobuf')), 'RC-B2 request evidence missing.');
  ensure(evidence.some((line) => line.includes('codec-command|codec=protobuf')), 'RC-B2 command evidence missing.');
  console.log('scenario RC-B2 passed');
}

File: framework/languages/dotnet/e2e/RegistrationCodec/Client/Scenarios/RcB2ProtobufCodecScenario.cs
// Verifies RC-B2 Typed Protobuf And MessagePack Codec behavior.
using RegistrationCodec.Client.Support;
using Zlink.HttpClient;

namespace RegistrationCodec.Client.Scenarios;

// RC-B2 verifies typed Protobuf and MessagePack codec round trips.
internal static class RcB2ProtobufCodecScenario
{
    public static async Task RunAsync(ZLinkHttpClient requester)
    {
        var result = (await requester.Post("/codec/roundtrip").Async<CodecScenarioRes>()).Body;
        ZlinkStreamAssert.Ensure(result.ProtobufValue.Contains("echo:rc-b2", StringComparison.Ordinal),
            "RC-B2 Protobuf reply mismatch.");
        ZlinkStreamAssert.Ensure(
            result.ProtobufValue.Contains("content:application/x-protobuf", StringComparison.Ordinal),
            "RC-B2 Protobuf content type mismatch.");
        ZlinkStreamAssert.Ensure(
            result.MessagePackValue.Contains("echo:rc-b3", StringComparison.Ordinal),
            "RC-B2 MessagePack reply mismatch.");
        ZlinkStreamAssert.Ensure(
            result.MessagePackValue.Contains("content:application/x-msgpack", StringComparison.Ordinal),
            "RC-B2 MessagePack content type mismatch.");

        Console.WriteLine("scenario RC-B2 passed");
    }
}
